openhantek/src/mainwindow.cpp

```diff
@@ -47,250 +35,84 @@ OpenHantekMainWindow::OpenHantekMainWindow(HantekDsoControl *dsoControl, DataAna
-    oscilloscopeToolBar = CreateToolBar(tr("Oscilloscope"));
-    oscilloscopeToolBar->addAction(startStopAction);
-
-    // View
-    viewToolBar = CreateToolBar(tr("View"));
-    viewToolBar->addAction(digitalPhosphorAction);
-    viewToolBar->addAction(zoomAction);
-}
+    restoreGeometry(settings->mainWindowGeometry);
+    restoreState(settings->mainWindowState);
 
-void OpenHantekMainWindow::addManualCommandEdit() {
     // Command field inside the status bar
-    commandEdit = new QLineEdit();
+    QLineEdit* commandEdit = new QLineEdit(this);
     commandEdit->hide();
 
-    commandAction = new QAction(tr("Send command"), this);
-    commandAction->setShortcut(tr("Shift+C"));
-
     statusBar()->addPermanentWidget(commandEdit, 1);
 
-    oscilloscopeMenu->addSeparator();
-    oscilloscopeMenu->addAction(commandAction);
-
-    connect(commandAction, &QAction::triggered, [this]() {
-        commandEdit->show();
-        commandEdit->setFocus();
+    connect(ui->actionManualCommand, &QAction::toggled, [this, commandEdit](bool checked) {
+        commandEdit->setVisible(checked);
+        if (checked)
+            commandEdit->setFocus();
     });
-    connect(commandEdit, &QLineEdit::returnPressed, [this]() {
-        Dso::ErrorCode errorCode = dsoControl->stringCommand(commandEdit->text());
 
-        commandEdit->hide();
+    connect(commandEdit, &QLineEdit::returnPressed, [this, commandEdit]() {
+        Dso::ErrorCode errorCode = this->dsoControl->stringCommand(commandEdit->text());
         commandEdit->clear();
-
+        this->ui->actionManualCommand->setChecked(false);
         if (errorCode != Dso::ErrorCode::NONE) statusBar()->showMessage(tr("Invalid command"), 3000);
     });
-}
 
-/// \brief Create all docking windows.
-void OpenHantekMainWindow::createDockWindows() {
-    registerDockMetaTypes();
-    horizontalDock = new HorizontalDock(settings, this);
-    triggerDock = new TriggerDock(settings, dsoControl->getSpecialTriggerSources(), this);
-    spectrumDock = new SpectrumDock(settings, this);
-    voltageDock = new VoltageDock(settings, this);
-}
-
-/// \brief Connect general signals and device management signals.
-void OpenHantekMainWindow::connectSignals() {
     // Connect general signals
-    connect(this, &OpenHantekMainWindow::settingsChanged, this, &OpenHantekMainWindow::applySettings);
     connect(dsoControl, &HantekDsoControl::statusMessage, statusBar(), &QStatusBar::showMessage);
 
     // Connect signals to DSO controller and widget
-    connect(horizontalDock, &HorizontalDock::samplerateChanged, this, &OpenHantekMainWindow::samplerateSelected);
-    connect(horizontalDock, &HorizontalDock::timebaseChanged, this, &OpenHantekMainWindow::timebaseSelected);
+    connect(horizontalDock, &HorizontalDock::samplerateChanged, [this]() {
+        this->dsoControl->setSamplerate(this->settings->scope.horizontal.samplerate);
+        this->dsoWidget->updateSamplerate(this->settings->scope.horizontal.samplerate);
```

Comment:
Do we really need all these detailed updateXXX() methods on DsoWidget side? We should probably emit MainWindow::settingsChanged() signal connected to applySettings() slots of DsoWidget and all dock windows.

Reason behind that is quite simple: DsoControl should be allowed to modify several attributes of settings object, hence multiple widgets might require update after.

Reply:
I think it's a misconception in the first place that the main window is responsible to keep DsoWidget and DsoControl in sync. And yes, the granularity is maybe to high. Lesser signals would be alright.

This PR is just about tidying up the Docks and converting the MainWindow though.